### ISSUE TITLE
Address boundary error caused by BatchNorm

### ISSUE THREAD

**Issue by cgarciae** [tensorflow/swift#169]
I used the MNIST demo code and added my own architecture. I successfully trained a model by using `Conv2D` layers only, but when I tried to add a `BatchNorm` layer I got this error:

```
Stack dump:
0.      Program arguments: /home/cristian/swift/tf-0.2/usr/bin/swift -frontend -interpret main.swift -disable-objc-interop -color-diagnostics -module-name main 
/home/cristian/swift/tf-0.2/usr/bin/swift[0x433e6d4]
/home/cristian/swift/tf-0.2/usr/bin/swift[0x433c45e]
/home/cristian/swift/tf-0.2/usr/bin/swift[0x433e892]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x12890)[0x7f5983b04890]
[0x7f595a037b23]
[0x7f595a034364]
/home/cristian/swift/tf-0.2/usr/bin/swift[0xc7c369]
/home/cristian/swift/tf-0.2/usr/bin/swift[0xc80420]
/home/cristian/swift/tf-0.2/usr/bin/swift[0x4efc90]
/home/cristian/swift/tf-0.2/usr/bin/swift[0x4c5d59]
/home/cristian/swift/tf-0.2/usr/bin/swift[0x4c1609]
/home/cristian/swift/tf-0.2/usr/bin/swift[0x4721c0]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f5968560b97]
/home/cristian/swift/tf-0.2/usr/bin/swift[0x4705da]
fish: “swift main.swift” terminated by signal SIGSEGV (Address boundary error)
```

Here is my code with the error, if you comment the 3 lines where variable `bn1` appears it will run properly:

```swift
import TensorFlow
import Python

let np = Python.import("numpy")

/// Reads a file into an array of bytes.
func readFile(_ filename: String) -> [UInt8] {
    let d = Python.open(filename, "rb").read()
    return Array(numpy: np.frombuffer(d, dtype: np.uint8))!
}

/// Reads MNIST images and labels from specified file paths.
func readMNIST(imagesFile: String, labelsFile: String) -> (images: Tensor<Float>,
                                                           labels: Tensor<Int32>) {
    print("Reading data.")
    let images = readFile(imagesFile).dropFirst(16).map { Float($0) }
    let labels = readFile(labelsFile).dropFirst(8).map { Int32($0) }
    let rowCount = Int32(labels.count)
    let columnCount = Int32(images.count) / rowCount

    print("Constructing data tensors.")
    return (
        images: Tensor(shape: [rowCount, columnCount], scalars: images) / 255,
        labels: Tensor(labels)
    )
}

// Ported from github.com/keras-team/keras/blob/master/examples/cifar10_cnn.py
public struct AllCNN: Layer {
    var conv1: Conv2D<Float>
    var bn1: BatchNorm<Float>
    var conv2: Conv2D<Float>
    var conv4: Conv2D<Float>
    
    public init(){
        self.conv1 = Conv2D<Float>(filterShape: (3, 3, 1, 32), strides: (2, 2), padding: .same)
        self.bn1 = BatchNorm<Float>(featureCount: 32)
        self.conv2 = Conv2D<Float>(filterShape: (3, 3, 32, 64), activation: relu)
        self.conv4 = Conv2D<Float>(filterShape: (3, 3, 64, 10))
    }
    

    @differentiable(wrt: (self, input))
    public func applied(to input: Tensor<Float>, in context: Context) -> Tensor<Float> {
        var net = input

        net = conv1.applied(to: net, in: context)
        net = bn1.applied(to: net, in: context)
        net = conv2.applied(to: net, in: context)
        net = conv4.applied(to: net, in: context)
        net = net.mean(alongAxes: [1, 2]).reshaped(toShape: [-1, 10])

        return net
    }
}

let epochCount = 20
let batchSize = 20

var bn1 = BatchNorm<Float>(featureCount: 32)

func minibatch<Scalar>(in x: Tensor<Scalar>, at index: Int) -> Tensor<Scalar> {
    let start = Int32(index * batchSize)
    return x[start..<start+Int32(batchSize)]
}

print("Before Init")

var classifier = AllCNN()

print("After Init")

let (images, numericLabels) = readMNIST(imagesFile: "data/train-images-idx3-ubyte",
                                        labelsFile: "data/train-labels-idx1-ubyte")
let labels = Tensor<Float>(oneHotAtIndices: numericLabels, depth: 10)


let context = Context(learningPhase: .training)
let optimizer = Adam<AllCNN, Float>(learningRate: 0.001)



// The training loop.
for epoch in 0..<epochCount {
    var correctGuessCount = 0
    var totalGuessCount = 0
    var totalLoss: Float = 0
    for i in 0 ..< Int(labels.shape[0]) / batchSize {
        var x = minibatch(in: images, at: i)
        x = x.reshaped(to: [Int32(batchSize), 28, 28, 1])
        let y = minibatch(in: numericLabels, at: i)
        // Compute the gradient with respect to the model.
        let grad_model = classifier.gradient { classifier -> Tensor<Float> in
            let ŷ = classifier.applied(to: x, in: context)
            let correctPredictions = ŷ.argmax(squeezingAxis: 1) .== y
            correctGuessCount += Int(Tensor<Int32>(correctPredictions).sum().scalarized())
            totalGuessCount += batchSize
            let loss = softmaxCrossEntropy(logits: ŷ, labels: y)
            totalLoss += loss.scalarized()
            return loss
        }
        // Update the model's differentiable variables along the gradient vector.
        optimizer.update(&classifier.allDifferentiableVariables, along: grad_model)

    }

    let accuracy = Float(correctGuessCount) / Float(totalGuessCount)
    print("""
          [Epoch \(epoch)] \
          Loss: \(totalLoss), \
          Accuracy: \(correctGuessCount)/\(totalGuessCount) (\(accuracy))
          """)
}
```
I am using Ubuntu 18.04, cuda 10 and cudnn 7.5, Swift for Tensorflow 0.2.

**Comment by cgarciae** [comment]
I tested this code on colab and it works.

**Comment by cgarciae** [comment]
Apparently the error only occurs when calling it like a script, I compiled it using `swiftc` and ran the binary and it worked.
